### PR TITLE
Some minor changes

### PR DESCRIPTION
**Pull request by naquad**
Hi.

Thank you for nvim-ipy, it saved a lot of time for me. I've made a number of additions to make it even better:

1. Ensure visual range is correct so one can run bottom-up selection.
2. I don't always have kernel running or QtConsole aside and I wanted IPython to start kernel if there's no connected. Not sure about this one, had couple of glitches, but mostly works fine.
3. Because I run separate QtConsole I wanted to hide IPython window (`IPython --no-window`) and be
4. able to append `\n` to executed code for better view in other console (`g:nvim_ipy_append` variable, default to empty string).
5. Treat end of file as end of cell option (`g:nvim_ipy_cell_eof`) so one doesn't have to close last cell explicitly.

Thank you!

### CODE REVIEW

File: rplugin/python3/nvim_ipy/__init__.py [hunk 1]
```diff
@@ -317,7 +337,7 @@ def ipy_run(self, args):
                     self.km.start_kernel(**self.km._launch_args)
             return
 
-        reply = self.waitfor(self.kc.execute(code,silent=silent))
+        reply = self.waitfor(self.kc.execute(code + self.append,silent=silent))
```
Comment (bfredl) [inline rplugin/python3/nvim_ipy/__init__.py]: does other clients (`jypyter console` is my reference) always make sure to end code in a `\n`? then we could do this unconditionally, without needing an option. (though I would add this to the line and visual run code, not here)

File: plugin/ipy.vim [hunk 2]
```diff
@@ -32,6 +32,11 @@ endfunction
 function! s:get_visual_selection()
     let [lnum1, col1] = getpos("'<")[1:2]
     let [lnum2, col2] = getpos("'>")[1:2]
+
+    if lnum1 > lnum2
```
Comment (bfredl) [inline plugin/ipy.vim]: Good catch!

**Comment by bfredl** [comment]
> Treat end of file as end of cell option (g:nvim_ipy_cell_eof) so one doesn't have to close last cell explicitly.

I would assume this is for the case with no separate cell start and end markers, ie `##` in python? I was thinking, if a string is supplied and not a list, then both the start of the buffer and the end of the buffer could implicitly be cells, without the need for a separate option.

File: rplugin/python3/nvim_ipy/__init__.py [hunk 3]
```diff
@@ -296,18 +302,32 @@ def ignore(self, msg_id):
...
+            Async(self).connect(args)
+        else:
+            self.connect(args)
 
     @neovim.function("IPyRun")
     def ipy_run(self, args):
         code = args[0]
         silent = bool(args[1]) if len(args) > 1 else False
+
+        if self.km is None:
```
Comment (bfredl) [inline rplugin/python3/nvim_ipy/__init__.py]: I would argue against this, it muddles the interface needlessly. But I would support a change that makes possible for an exernal mapping to do this. Maybe a variable with connection status, and `User` autocmd when connected would be useful?

File: rplugin/python3/nvim_ipy/__init__.py (hunk 1 above)
Comment (naquad) [reply rplugin/python3/nvim_ipy/__init__.py]: No, it is purely QtConsole thing.

File: rplugin/python3/nvim_ipy/__init__.py (hunk 3 above)
Comment (naquad) [reply rplugin/python3/nvim_ipy/__init__.py]: Maybe lets then just make a variable making this behavior optional? Sort of `g:nvim_ipy_autostart` with default = 0.

**Comment by bfredl** [comment]
I cherry-picked some changes and made some others myself: Beginning and end of buffer should now work as cells if single separator string is used. Interactive bindings will always make sure that input ends with `\n` to be consistent with jupyter consoles.

Thanks for you contributions!

**Pull request closed without merging**